src/bin/head.rs

```diff
@@ -0,0 +1,137 @@
+#![deny(warnings)]
+
+extern crate coreutils;
+
+use std::env;
+use std::fs;
+use std::io::{self, BufRead, Read, Write};
+use coreutils::extra::{OptionalExt, fail};
+
+static MAN_PAGE: &'static str = r#"NAME
+    head - output the first part of a file
+
+SYNOPSIS
+    head [[-h | --help] | [-n LINES] | [-c BYTES]] [FILE ...]
+
+DESCRIPTION
+    Print the first 10 lines of each FILE to standard output. If there are no files, read the standard input. If there are multiple files, prefix each one with a header containing it's name.
+
+OPTIONS
+    -h
+    --help
+        Print this manual page.
+
+    -n [-]LINES
+        Print the first LINES lines. If prefixed with a minus, print all but the last LINES lines.
+
+    -c [-]BYTES
+        Print the first BYTES bytes. If prefixed with a minus, print all but the last BYTES bytes.
+
+AUTHOR
+    Written by Žad Deljkić.
+"#;
+
+// lines - true if outputing lines, false if outputing bytes
+// num - number of lines/bytes specified
+// skip - false if outputing first num lines/bytes, true if outputing all but the last num lines/bytes (i.e. skip the last num lines/bytes)
+#[derive(Copy, Clone)]
+struct Options {
+    lines: bool,
+    num: usize,
+    skip: bool,
+}
+
+// get the last line/byte up to which we read
+// while taking care to stay within bounds
+fn get_last(num: usize, len: usize, skip: bool) -> usize {
+    if skip {
+        if num <= len {
+            len - num
+        } else {
+            0
+        }
+    } else {
+        if num <= len {
+            num
+        } else {
+            len
+        }
+    }
+}
+
+fn head<R: Read>(mut input: R, opts: Options) -> io::Result<()> {
+    if opts.lines {
+        let lines: Vec<String> = try!(io::BufReader::new(input).lines().collect());
```

Comment:
Why collect the whole buffer of lines? That seems kinda slow.

GNU head will simply read the N first lines.